sirepo/jupyterhub.py

```diff
@@ -32,33 +35,60 @@ def __init__(self, *args, **kwargs):
         sirepo.server.init()
 
     async def authenticate(self, handler, data):
-        d = self._check_permissions(handler)
-        if 'username' in d:
-            return d.username
-        elif 'uri' in d:
-            handler.redirect(d.uri)
-            raise tornado.web.Finish()
         # returning None means the user is forbidden (403)
         # https://jupyterhub.readthedocs.io/en/stable/api/auth.html#jupyterhub.auth.Authenticator.authenticate
-        return None
+        return self._check_permissions(handler).get('username')
 
     async def refresh_user(self, user, handler=None):
-        try:
-            return bool(self._check_permissions(handler).get('username'))
-        except Exception:
-            # Returning False is what the jupyterhub API expects and jupyterhub
-            # will handle re-authenticating the user.
-            # https://jupyterhub.readthedocs.io/en/stable/api/auth.html#jupyterhub.auth.Authenticator.refresh_user
-            return False
-        raise AssertionError('should not get here')
+        # Reading jupyterhub code the handler is never None
+        # We need the handler for cookies and redirects
+        assert handler, \
+            'handler should never be none'
+        # Returning True/False is what the jupyterhub API expects and jupyterhub
+        # will handle re-authenticating the user if needed.
+        # https://jupyterhub.readthedocs.io/en/stable/api/auth.html#jupyterhub.auth.Authenticator.refresh_user
+        return bool(self._check_permissions(handler).get('username'))
 
     def _check_permissions(self, handler):
+        def _cookies(response):
+            for k, v in response.cookies.get_dict().items():
+                handler.set_cookie(k, v)
+
+
+        def _maybe_html_redirect(response):
+            m = re.search(r'window.location = "(.*)"', pkcompat.from_bytes(response.content))
+            m and self._redirect(handler, f'{self.sirepo_uri}{m.group(1)}')
```

Comment:
Does there need to be a sirepo_uri? Doesn't the redirect just work on `/` without http://....?